src/compiler/checker.ts

```diff
@@ -11225,6 +11234,56 @@ namespace ts {
             }
         }
 
+        /**
+         * Determines if all declarations in the given list are function declarations, and that
+         * all the declarations have the same name. This is useful for how we report errors.
+         *
+         * For instance, if a user has
+         *
+         *      export default function foo() { }
+         *      export default function bar() { }
+         *
+         * We want to tell the user that there are multiple default exports. However, if a user has
+         *
+         *      export default function foo() { }
+         *      export default function foo() { }
+         *
+         * It is probably more accurate to say that there are duplicate function implementations.
+         * 
+         * @param declarations
+         */
+        function areAllFunctionDeclarationsWhereSomeNamesDiffer(declarations: Declaration[]) {
+            let numDeclarations = declarations.length;
+            Debug.assert(numDeclarations > 0, "Expected at least one declaration.");
+
+            let firstDeclaration = declarations[0];
+            if (firstDeclaration.kind !== SyntaxKind.FunctionDeclaration) {
+                return false;
+            }
+
+            let firstName = (firstDeclaration as FunctionDeclaration).name;
+            for (let i = 1; i < numDeclarations; i++) {
+                let currentDeclaration = declarations[i];
+                if (currentDeclaration.kind !== SyntaxKind.FunctionDeclaration) {
+                    return false;
+                }
+
+                let currentName = (currentDeclaration as FunctionDeclaration).name;
+                if (currentName) {
+                    if (!firstName || currentName.text !== firstName.text) {
+                        // Either 'firstName' must be undefined or the text must differ.
+                        return true;
+                    }
+                }
+                else if (firstName) {
+                    // 'currentName' is not defined but 'firstName' is.
+                    return true;
+                }
+            }
+
+            return false;
+        }
+

```

Comment:
Good point. Since we're reporting an error here, perf isn't exactly a concern anyway.

Also, as a heads up, ES5 array functions such as `some` and `every` aren't permitted in the compiler because we sort of support running on ES3 runtimes.